create test report generation durring CI: fix #166

bebuch opened the pull request:
I break down the description of the changes by file. ;-)

### `.circleci/config.yml`

This is the primary change. As requested in the issue, the creation of the report is now tested in addition to the unit tests. Based on this, I can introduce more such 'syntax tests' later.

I have removed the comments from circleci's sample file. I take the position that comments should never be used in source code to explain the programming language. It makes the source code confusing and the explanation is practically always incomplete and/or outdated. Instead I have linked the doc of the circleci YAML dialect at the top.

The `orbs` definition is thrown out, because we do not use that.

I have defined three jobs:

1. `run-tests` installs our `tests` package including its dependencies and then runs the tests. This is identical to the previous pipeline.
2. `run-examples` installs our `emva1288` package including its dependencies and runs `doc/examples/sample_report.py` to generate the tex version of the report. The generated directory is saved. In principle, other samples can also be executed in this job to check them for syntax errors.
3. `compile-report` is executed in the current Docker container of the `texlive` distribution. It loads the saved report folder and uses `latexmk` to generate the PDF file. Like the usual IDEs, `latexmk` repeatedly calls `pdflatex` until the final PDF is generated.

The `unittests` workflow calls the `run-tests` job. The `create-report` workflow calls the `run-examples` and `compile-report` jobs. Thereby `compile-report` has to wait until `run-examples` has finished successfully.

### `doc/examples/sample_report.py`

I reduced the image resolution for the sample report generation from 640x480 to 160x120.

The 640x480 requires 13.4 GB of RAM. I don't want to impose that on the CI or the end user in an example. In both cases we don't know how potent the hardware is and the examples should work by default. The 160x120 still requires 900 MB of RAM.

If I find time, I'll take a look to see if that can be reduced.

If reducing the resolution causes problems with the content, please let me know.

### `emva1288/report/templates/op.tex`

That's the bug @zhaoyutong43210 already mentioned in [#166](https://github.com/EMVA1288/emva1288/pull/163#issuecomment-1243905583). This is exactly the kind of bug the new pipeline is supposed to detect, and it did. ;-)

### `emva1288/report/templates/report.tex`

I have replaced the `xwatermark` latex package with `draftwatermark`. `xwatermark` does not work with most Tex distributions after 2020 and should not be used anymore. (See https://tex.stackexchange.com/questions/634973/extra-endgroup-begindocument-frustrations#answer-635421)

bebuch commented:
Looks like the CI doesn't run when you change its config. Here is the run from my Repo:

- https://app.circleci.com/pipelines/github/bebuch/emva1288/11

aturcotte commented:
I always try generating a PDF report when reviewing pull requests. I tried generating one and received this error:
![image](https://user-images.githubusercontent.com/862398/190215486-b3e1d477-d7b2-412b-99aa-e8874339eb43.png)

So I checked out master, since this previously worked, but I receive a different error:
![image](https://user-images.githubusercontent.com/862398/190215938-6880aa2d-46e3-49ff-b03d-7936162a14ad.png)

Then I checked out v0.6.3 and confirmed that it works fine with that version. So either a regression was introduced recently or there is something on my side preventing it from working.

@bebuch Are you able to produce a PDF report locally?

bebuch commented:
The second one is a regression in `emva1288/report/templates/op.tex` introduce 7 days ago by [e6e9b836](https://github.com/EMVA1288/emva1288/commit/e6e9b836d7656719c51b7679209a0c740d98be72#diff-539a798adb488b7696f464c0a3d16443a1df0371c1afa375bb355c284d930371R109). It's fix by this PR.

What Tex-Distribution and version do you use? Need to know to reproduce.

I locally run anything in the same docker containers as the CI, so I don't have any "local" setup. ;-)

aturcotte commented:
I am using texlive 2019.20200218-1 on Ubuntu 20.04.

aturcotte commented:
Apologies. I forgot to add the `[doc]` extra when installing the package. It works fine.

aturcotte commented on .circleci/config.yml:
```diff
@@ -1,47 +1,50 @@
+    steps:
+      - checkout
+      - run:
+          command: pip install --upgrade pip; pip install .[emva1288]
```
The `[emva1288]` extra is invalid. I'm guessing it should be `[doc]`?
```suggestion
          command: pip install --upgrade pip; pip install .[doc]
```

bebuch replied:
Yep, I haven't looked up how the installation process works under Python yet and just guessed. Unfortunately guessed wrong. ^^'

That there is a warning I had overlooked so far. From skimming, these 'extras' seem to be defined in `setup.cfg`. I will have a look on Friday how this works. Thanks for pointing it out!

Since it works with wrong value the line can be gone completely.

But I will adjust the process again, because I want to test the package building with. The config file looks suspiciously like it defines some kind of Python package. ;-)

bebuch commented:
@aturcotte After some reading and a night of sleep I decided to commit your suggestion here. I will do a separate PR for my further planned changes. ;-)

fariza commented on doc/examples/sample_report.py:
```diff
@@ -8,8 +8,8 @@
 
-dataset_generator = DatasetGenerator(width=640,
-                                     height=480,
+dataset_generator = DatasetGenerator(width=160,
```
I understand that for tests we need smaller images so it generates fast.
I'm not fan of callable python files, but in this case, I would prefer that we add a `if __name__ == __main__` and use `arparse` to change the width and height for the tests. 
What do you think?

fariza replied:
just to clarify why I want to keep relatively large images.
When developing the report, smaller image give strange spatial graphics, making it harder to differentiate between report look from actual processing problem

bebuch replied:
I think a clear and easy to understand example is a priority. If you think larger images are necessary, I would suggest to adjust the file via sed in the CI.

fariza replied:
I don't understand why using sed is easier than just creating an entry point?

bebuch replied:
I have only a very rudimentary knowledge of Python and its concepts so far. But as far as I understand it, this variant would have to extend the Python script doc example by something that is not relevant in the context of the doc. If you look at it only from the point of view of a documentation reader, it is easier to leave out the entry point stuff. A `if __name__ == '__main__'` will throw up the question of why this is necessary at this point. In the implementation of the library, I would be okay with such an extension, but in the documentation I think it's best to focus entirely on the relevant content.

Overall, the solution with sed is of course a bit more complex, but this complexity is encapsulated in the CI script and thus in a place that hardly any user of the repo will come into contact with.

bebuch replied:
I thought again about the `if __name__ == '__main__'` variant. We would call the actual script function in the `if __name__ == '__main__'` branch. In the CI we include the script via another file, which means that the `if __name__ == '__main__'` branch is not executed. Consequently, we can call the script function with its own parameters.

If we set up all examples this way, I think this is okay as "code overhead". Especially since this seems to be a common pattern in Python, and moving the code from the global namespace to that of a function has a few more advantages. (Have read a bit about this in the meantime).

Only downside would be that we wouldn't test the `if __name__ == '__main__'` branch within the CI, of course. So if we pass invalid parameters in this branch, users will get an error message when running the example script, but the CI won't get anything out of it.

Should I adjust the PR accordingly? (The `if __name__ == '__main__'` I only include in the report for now. The adaptation of the other examples would be done in a separate PR).

fariza replied:
@bebuch yes, let's adapt this one only

bebuch commented:
I have modified the report example a bit, so that the parameters for the CI can be easily replaced and at the same time all duplicated values are defined as variables at the beginning. I think this is also better for the doc.

The CI uses `sed` to adjust the necessary lines before executing the script. The `sed` expression fails if the searched line is not found or is found multiple times.

The replacement is noted in the log (command line output) with a green "success" or a red "fail". In case of a "fail" the pipline also fails. The changed file will be output again completely in the log after the changes, to be able to debug easily if necessary. Afterwards the execution takes place as before.

So I'm ready for the next round of reviews :-)

fariza commented:
@zhaoyutong43210 please review

bebuch commented:
I have added the changes suggested by @fariza.

I moved the dataset generation to the `if __name__ == '__main__'` branch, as commented out code also provides the ability to load. As a result, there is now more than one line of code in the area that is not executed by the CI. For now, I would leave it at that. Perhaps there will be a better variant in the future.

@aturcotte @zhaoyutong43210 Ready for review again.

fariza commented:
@zhaoyutong43210 @aturcotte please review
On my side I'm ok with the solution for the parameters on the test and on the sample

aturcotte commented on doc/examples/sample_report.py:
```diff
@@ -1,72 +1,92 @@
-parser = process.ParseEmvaDescriptorFile(fname)
-imgs = process.LoadImageData(parser.images)
-dat = process.Data1288(imgs.data)
 
```
There is an extra blank line here. Otherwise, this looks good.
```suggestion
```

aturcotte commented on doc/examples/sample_report.py:
```diff
@@ -1,72 +1,92 @@
 from emva1288 import report
-from emva1288.camera.dataset_generator import DatasetGenerator
 import os
 
```
Put an extra blank line here.
```suggestion


```

aturcotte commented on doc/examples/sample_report.py:
```diff
@@ -1,72 +1,92 @@
-                             basic=basic)
+    # Generate the latex files
+    myreport.latex()
 
```
Put an extra blank line here.
```suggestion


```

fariza commented:
@zhaoyutong43210 please review

zhaoyutong43210 commented:
this looks good to me, I have been watching this, but I didn't add comments since I didn't have time to try it.
